### PR TITLE
feat: transaction_ignore_urls wildcard matching

### PR DESCRIPTION
**Pull request by astorm**
Will Fix: elastic/apm-agent-nodejs#1689

Implements `ELASTIC_TRANSACTION_IGNORE_URL`/`transactionIgnoreUrls` configuration for wildcard URL ignoring.  The intent of this new configuration is to gives users the ability to skip generating transactions for a request to a URL that matches an exact string, or a URL that matches simple wildcard pattern.  This wildcard pattern will behave consistently across language agents.

This PR leaves the existing `ignoreUrls` configuration in place to allow users access to more powerful regular expression matching _and_ to preserve existing functionality for users using this older configuration setting. 

This PR also introduces a new dependency, [`escape-string-regexp`](https://www.npmjs.com/package/escape-string-regexp).  In order to implement the suggested wildcard matching algorithm we needed the ability to escape a string for use in a regular expression.  Since Node.js doesn't have this functionality natively, I pulled in `escape-string-regexp` rather than attempt to write (easy) and maintain (harder) this functionality ourselves.  Happy to revisit this decision if we'd rather implement this functionality ourselves. 

### Open items for consideration/completion
- [x] is `escape-string-regexp` an acceptable dependency for ignore regular expressions?
- [x] Docs
- [x] Fixture downloading
- [x] central config handling? How does this work in Node Agent, do we need more here?
- [x] should leading `/` be assumed? (we'll prepend this, meaning a configured value of 'foo*' and '/foo*' will match the same)
- [x] need clarity on `(?-i)` and `(?+i)` prefixes across agents
- [x] Does this apply only to transactions? (yes, `isRequestBlacklisted` is only called in `instrumentRequest`, and `instrumentRequest` is only used when wrapping http(s).Server's `emit` method (i.e. it's used in transaction starting)

### Checklist

- [x] Implement code
- [x] Add tests
- [x] Update TypeScript typings
- [x] Update documentation
- [x] Commit message follows [commit guidelines](https://github.com/elastic/apm-agent-nodejs/blob/master/CONTRIBUTING.md#commit-message-guidelines)

### CODE REVIEW

File: lib/config.js
Comment (astorm) [inline lib/config.js:386]: Move `normalizeIgnoreOptions` down in the list in order to ensure `normalizeArrays` had run, and that the `transactionIgnoreUrls` value would already be converted from a comma,delimited,string into an array (if necessary)

File: docs/configuration.asciidoc
Comment (bmorelli25) [inline docs/configuration.asciidoc:217]:
```suggestion
    '(?+i)/caseSensitiveSearch'
```

**Review by bmorelli25 — COMMENTED** [review]
Nice! One small typo, otherwise [docs](https://apm-agent-nodejs_1876.docs-preview.app.elstc.co/guide/en/apm/agent/nodejs/master/configuration.html#transaction-ignore-urls) LGTM 🎉

**Comment by astorm** [comment]
jenkins run lint

File: lib/config.js
Comment (trentm) [inline lib/config.js:407]: I don't think we should do this "favour". 

- I suspect other agents don't do it, which would slightly/subtly break commonality. The Python agent at least doesn't do this.
- Combined with a possible empty string (which could come from a dangling comma like `ELASTIC_TRANSACTION_IGNORE_URLS=a,b,`) this would result in a pattern to ignore the root URL, `'/'`, which would be a surprise.

File: lib/instrumentation/http-shared.js
Comment (trentm) [inline lib/instrumentation/http-shared.js:83]: nit: restore the blank line before the `ua` block?

File: lib/wildcard-matcher.js
Comment (trentm) [inline lib/wildcard-matcher.js:1]: Should we `'use strict'` at the top of all our JS files?

nit: A relatively brief top-comment at the top of each file giving its general intent would be helpful. E.g. could mention this implements a common "WildcardMatcher" spec from apm.git (though that repo doesn't have a good link for that spec, just the JSON file of tests).

File: test/tmp-fixtures.js
Comment (trentm) [inline test/tmp-fixtures.js:1]: Should this be dropped now that test/fixtures/json-specs/wildcard_matcher_tests.json is available?

File: test/wildcard-matcher.js
Comment (trentm) [inline test/wildcard-matcher.js:1]: `'use strict'` ?

File: test/config.js
Comment (trentm) [inline test/config.js:475]: Would want to drop this test if dropping the "auto prefix with '/'" feature above.

File: test/central-config-enabled.js
Comment (trentm) [inline test/central-config-enabled.js:29]: Drop this commented line?

File: lib/wildcard-matcher.js
Comment (trentm) [inline lib/wildcard-matcher.js:33]: The spec doesn't mention `(?+i)` at all. I think you could simplify to:

```
if (pattern.startsWith('(?-i)')) {
    opts = []     // or perhaps change `opts` to be a specific caseSensitive bool instead of generic regexp flags
    pattern = pattern.slice(5)
}
```

and drop `parsePatternAndOptsFromPattern`.

File: docs/configuration.asciidoc
Comment (trentm) [inline docs/configuration.asciidoc:202]: It took me a LONG time to realize that this is case-*insensitive* by default.  

1. Eww, but that's the way it was specified.
2. I think we should have a clear "Matching is case-sensitive by default. Prefix your pattern with `(?-i)` to make the matching case-sensitive."
3. Do you really want to extend the spec by supporting `(?+i)`? At least the Python agent doesn't support that. I argue lower in this review that the code can be made simpler by only supporting `(?-i)` and not `(?+i)`.

File: docs/configuration.asciidoc
Comment (trentm) [inline docs/configuration.asciidoc:217]:
```suggestion
    '(?-i)/caseSensitiveSearch'
```

**Review by trentm — CHANGES_REQUESTED** [review]

File: test/wildcard-matcher.js
Comment (astorm) [reply test/wildcard-matcher.js:2]: _nod_ `use strict` at the top of every JS file should be the defacto choice.  I'm used to ESLint catching this -- maybe another bit of evidence to move away from `standard` or at least enhance our CI it with some additional ESLint rules

File: lib/wildcard-matcher.js
Comment (astorm) [reply lib/wildcard-matcher.js:2]: _nod_ -- seems reasonable -- will add.

File: CHANGELOG.asciidoc
Comment (astorm) [reply CHANGELOG.asciidoc:15]: will change to `transactionIgnoreUrls`, and I'm up for giving spec links (when available) a try with our newer, more verbose changelog overlords :)   Linked to https://github.com/elastic/apm/issues/144, which is the closest we have to a spec for this feature.

File: docs/configuration.asciidoc
Comment (astorm) [reply docs/configuration.asciidoc:190]: Good call out.  If I'm reading that right "Dynamic" configuration means users can confg this via central config. 

How about we hold off this for now? My thinking is this configuration isn't available yet in the 7.10 release, and agents release off cycle from mainline elastic.  Once the `TRANSACTION_IGNORE_URL` setting makes it into Kibana/APM-UI proper we can come back and mark this as dynamic. 

Thoughts? (if so, just unresolved and we can keep talking)

File: docs/configuration.asciidoc
Comment (astorm) [reply docs/configuration.asciidoc:199]: 👍

File: lib/config.js
Comment (astorm) [reply lib/config.js:392]: 👍 Good call out -- `normalizeArrays` was a late discovery, will change.

File: lib/config.js
Comment (astorm) [reply lib/config.js:397]: 👍

File: docs/configuration.asciidoc
Comment (trentm) [reply docs/configuration.asciidoc:190]: @bmorelli25 Thoughts on this? Is there a way we could say "in APM server version 7.11" alongside the "dynamic configuration" badge/link?

File: docs/configuration.asciidoc
Comment (astorm) [reply docs/configuration.asciidoc:193]: Hmmm -- looks like there's precedent for calling out the comma separation in the `ELASTIC_APM_DISABLE_INSTRUMENTATIONS`, so I'm game.  The "no commas" is trickier -- it's only there as a side effect of the the way APM Server will be sending us that data AND URLs don't seem common in URL Paths.  On the flip side though -- truth :)   I've added a comment about commas.  I'll let @bmorelli25 be the final arbiter once we've got this locked in.

File: docs/configuration.asciidoc
Comment (astorm) [reply docs/configuration.asciidoc:207]: Agreed -- once this is actually available as a Kibana central config option we can/should come back and make mention of it.  

Similar to "Added in version x.y.z" -- that would be nice to have as a user -- on the flip side that's a burden on us to start adding when every configuration option was added.  I'm kind of 50/50 on this -- maybe table that until we know what the actual next version of the agent is? (something to add to our release check list maybe)

File: lib/config.js
Comment (astorm) [reply lib/config.js:407]: That's fair -- I can back off this one.  (Likely came from years of trauma with apache handling the path differently in the main config vs. `.htaccess` files :))

File: lib/instrumentation/http-shared.js
Comment (astorm) [reply lib/instrumentation/http-shared.js:83]: 👍

File: test/config.js
Comment (astorm) [reply test/config.js:475]: 👍

File: test/tmp-fixtures.js
Comment (astorm) [reply test/tmp-fixtures.js]: 🤦 👍

File: test/wildcard-matcher.js
Comment (astorm) [reply test/wildcard-matcher.js:2]: 👍

File: test/central-config-enabled.js
Comment (astorm) [reply test/central-config-enabled.js]: we regret the error -- we also regret the next occurrence of this error -- I can be comment blind towards the end of a feature 🤦

File: docs/configuration.asciidoc
Comment (astorm) [reply docs/configuration.asciidoc:202]: Hmm Re: `(?+i)` -- my memory is I saw that used in the Java Agent's wildcard implementation, and I included parsing it out in case someone was copying rules between agents. Going back now though I don't see it anywhere -- so good catch and a reminder for me meditate on the fallibility of memory  

i.e. I'll remove the code for handling the + case and be explicit in the language about what is and isn't supported. 

And yes -- I'm also looking forward to participating in spec discussions instead of implementing specs decided on before our time. :)

File: docs/configuration.asciidoc
Comment (trentm) [reply docs/configuration.asciidoc:207]: Personally I'd make mention of it now, referencing future "Kibana 7.11", but I'm also fine doing it later.

> maybe table that until we know what the actual next version of the agent is?

Sounds good.

File: lib/wildcard-matcher.js
Comment (astorm) [reply lib/wildcard-matcher.js:40]: agreed -- dropped the extra function in favor the single conditional.

File: docs/configuration.asciidoc
Comment (bmorelli25) [reply docs/configuration.asciidoc:190]: Good catch. Whether we wait on adding this badge to the docs or not, it's probably important either way to mention that stack version 7.11+ is required. I think text next to the badge is fine. We could add it to the badge too, but that might be overkill. . . 

Screen Shot 2020-11-18 at 3 52 58 PM

Comment (trentm) [reply docs/configuration.asciidoc:190]: @bmorelli25 Ooh, I like that. What's the asciidoc syntax for that?

Comment (bmorelli25) [reply docs/configuration.asciidoc:190]: Badges are a bit of a hack in our build as we don't have native support for them. In this case, the new [badge](https://img.shields.io/badge/Dynamic-Requires%20APM%20Server%207.11%2B-blue) needs to be added to the images directory and then referenced:

`<<dynamic-configuration, image:./images/dynamic-config-stack-711.svg[] >>`

I will say-- after thinking some more on this, I don't think it makes to add this badge to the docs yet; I think it will only confuse users.

Comment (astorm) [reply docs/configuration.asciidoc:190]: Thanks @bmorelli25 -- we've captured our intent to update this label once the feature lands in kibana: https://github.com/elastic/apm-agent-nodejs/issues/1883

File: lib/wildcard-matcher.js
Comment (trentm) [inline lib/wildcard-matcher.js:10]: nit:

```suggestion
// Converts elastic-wildcard pattern into a
```

**Review by trentm — APPROVED** [review]
Just a spelling typo in the docs and a typo. Soft approving.

**Comment by astorm** [comment]
Merging, and will handle the docs typos in a quick follow-on PR.

**Pull request merged**
